hidden/Models/GlobalEventMoniter.swift

```diff
@@ -0,0 +1,40 @@
+//
```

Comment:
I just add Local monitor for Hidden bar it self to by pass your current limitation. Please test your idea again with this update and let me know.
```

import Cocoa

protocol EventMoniter{
    func start()
    func stop()
}

class GlobalEventMoniter: EventMoniter {
    
    var monitor: Any?
    let mask: NSEvent.EventTypeMask
    let handler: (NSEvent) -> ()
    var isStarted : Bool {
        return self.monitor != nil
    }
    
    init(mask: NSEvent.EventTypeMask, handler: @escaping (NSEvent) -> ()) {
        self.mask = mask
        self.handler = handler
    }
    
    deinit {
        stop()
    }
    
    func start() {
        guard !isStarted else { return }
        self.monitor = NSEvent.addGlobalMonitorForEvents(matching: mask, handler: handler)
    }
    
    func stop() {
        guard isStarted else { return }
        NSEvent.removeMonitor(self.monitor!)
        self.monitor = nil
    }
    
}

class LocalEventMoniter: EventMoniter {
    
    var monitor: Any?
    let mask: NSEvent.EventTypeMask
    let handler: (NSEvent) -> (NSEvent?)
    var isStarted : Bool {
        return self.monitor != nil
    }
    
    init(mask: NSEvent.EventTypeMask, handler: @escaping (NSEvent) -> (NSEvent?)) {
        self.mask = mask
        self.handler = handler
    }
    
    deinit {
        stop()
    }
    
    func start() {
        guard !isStarted else { return }
        self.monitor = NSEvent.addLocalMonitorForEvents(matching: mask, handler: handler)
    }
    
    func stop() {
        guard isStarted else { return }
        NSEvent.removeMonitor(self.monitor!)
        self.monitor = nil
    }
    
}
```

```

import AppKit

class MouseMovementHandler {
    
    var expandHandler   : () -> Void
    var collapseHandler : () -> Void
    var isCollapsed     : () -> Bool
    
    var eventMoniters = [EventMoniter]()
    
    private var menubarMinY: CGFloat = .nan
    
    private var previousMouseLocation = NSEvent.mouseLocation
    
    private var interactedWithMenuBar = false
    private var mouseIsOnMenuBar      = false {
        didSet {
            guard mouseIsOnMenuBar != oldValue else { return }
            if mouseIsOnMenuBar {
                interactedWithMenuBar = false
                expandHandler()
            } else if !interactedWithMenuBar {
                collapseHandler()
            }
        }
    }
    
    init(expandHandler: @escaping ()->Void, collapseHandler: @escaping ()->Void, isCollapsed: @escaping ()->Bool) {
        self.expandHandler = expandHandler
        self.collapseHandler = collapseHandler
        self.isCollapsed = isCollapsed
        
        setupEventMoniter()
        updateAutoExpandPreferences()
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(updateAutoExpandPreferences),
            name: .prefsChanged,
            object: nil)
    }
    
    private func setupEventMoniter() {
        eventMoniters.append(GlobalEventMoniter(mask: .mouseMoved, handler: {[weak self] event in self?.mouseMoved(event)}))
        eventMoniters.append(GlobalEventMoniter(mask: .leftMouseDown, handler: {[weak self] event in self?.mouseClicked(event)}))
        eventMoniters.append(GlobalEventMoniter(mask: .rightMouseDown, handler: {[weak self] event in self?.mouseClicked(event)}))
        
        eventMoniters.append(LocalEventMoniter(mask: .mouseMoved, handler: mouseMoved))
        eventMoniters.append(LocalEventMoniter(mask: .leftMouseDown, handler: mouseClicked))
        eventMoniters.append(LocalEventMoniter(mask: .rightMouseDown, handler: mouseClicked))

    }
    
    @objc private func updateAutoExpandPreferences() {
        if Preferences.autoExpand {
            eventMoniters.forEach { $0.start() }
        } else {
            eventMoniters.forEach { $0.stop() }
        }
    }
    
    private func updateMenubarMinY() {
        // Cocoa uses top left coords, but CG uses bottom left coords
        // To avoid convertion everytime, menubarMinY is inverted on initialize
        guard let screenHeight = Util.screenWithMouse?.frame.height,
              let menubarFrame = WindowInfo.cgRectOfWindow(named: "Menubar")
        else {
            return
        }
        
        menubarMinY = screenHeight - menubarFrame.height
    }
    
    @discardableResult
    private func mouseMoved(_ event: NSEvent) -> NSEvent? {
        
        if !isCollapsed() && !(NSEvent.mouseLocation ~= previousMouseLocation) {
            updateMenubarMinY()
            previousMouseLocation = NSEvent.mouseLocation
        }
        
        if NSEvent.mouseLocation.y >= menubarMinY {
            self.mouseIsOnMenuBar = true
        } else {
            self.mouseIsOnMenuBar = false
        }
        return event
    }
    
    @discardableResult
    private func mouseClicked(_ event: NSEvent) -> NSEvent? {
        print(NSEvent.mouseLocation.y)
        if NSEvent.mouseLocation.y >= self.menubarMinY {
            
            self.interactedWithMenuBar = true
        } else {
            self.interactedWithMenuBar = false
        }
        return event
    }
    
}

```

Reply:
OK I didn't know NSEvent also has a local monitor thing. But I think that I would be much simpler to have a `EventMonitor` class that contains both NSEventLocalMoniter and GlobalMoniter, instead of two different classes.

Reply:
Sorry for the late reply, but I believe this is solved in the latest commit